Add ContentTypeForm functionality to DecodeForm

syntaqx opened the pull request:
Some of my companies legacy applications require the ability to use `ContentTypeForm`. I didn't realize go-chi/render didn't support this until we started running into random breakage..

syntaqx commented:
Literally I just want the ability for decoding the following:

```
$ curl -d "name=test" -X POST http://mysite.com/create
```

I would say we could spend the time to recreate the functionality of gorilla/schema if you're trying to keep this repository free of dependencies, but it's pretty much the defacto standard for dealing with form data, and has fixed a lot of edge cases around them.

VojtechVitek commented:
1. https://github.com/google/go-querystring
2. https://github.com/gorilla/schema
3. https://github.com/ajg/form

We have couple options. Seems like 1. is the most mature.

syntaqx commented:
If I'm not mistaken, [go-querystring](https://github.com/google/go-querystring) is specifically for the case of utilizing urlencoded query arguments, rather than extracting form data into url.Values.

When comparing [ajg/form](https://github.com/ajg/form) and [gorilla/schema](https://github.com/gorilla/schema), it seems like both implementations could be viable, it's honestly a flip of the coin on that one.

syntaqx commented:
@VojtechVitek How would you like me to proceed with this? I believe that gorilla/schema is a good choice for this, but if you would like to vet the others, that's reasonable.

VojtechVitek commented:
@syntaqx you're right, go-querystring only implements encoding.

We should evaluate decide between the two other packages. Both seem very mature to me.

However, https://github.com/ajg/form looks simpler and more elegant to me.

gorilla/schema looks quite complex for what it's supposed to do, especially the internal cache: https://github.com/gorilla/schema/blob/master/cache.go#L18-L33

syntaqx commented:
Sorry for the massive delay on this. I've switched to using `ajg/form`. Please let me know if you have additional concerns.

syntaqx commented:
@VojtechVitek Let me know if you need anything from me. Would love to switch over to using this package instead of a fork again!

VojtechVitek commented:
@syntaqx thanks, I'll take a look when I have some spare time. I'm too busy right now with my job

VojtechVitek commented on decoder.go:
```diff
@@ -7,6 +7,8 @@ import (
 	"io/ioutil"
 	"net/http"
+
+	"github.com/ajg/form"
```
should we bring in github.com/ajg/form as golang/dep dependency?

syntaqx replied:
I think that's a more global question for you, for the project. Currently there's no dependency tooling, so I didn't add any.

otraore commented:
Any reason as to why this isn't merged yet, we need this at work too.

VojtechVitek commented:
@otraore I want to get this merged together with golang/dep files.

syntaqx commented:
I've added dep.

syntaqx commented:
Any follow up on this?

VojtechVitek commented:
Sorry I'm too busy at this point. I'll try to come back to this when I have some spare time.

syntaqx commented:
Hey @VojtechVitek - I know you've been busy, but was just wondering if you've had the chance to look at this? If not, perhaps you'd be willing to add myself or someone else as a maintainer? Would love to be able to keep using this, but without support it's becoming very unreasonable.

syntaqx commented:
As an aside, `gorilla/schema` seems to be a better package overall, but I think the general purpose of this PR can use either. Seems that there are occasional reflection issues with `ajg/form` where it has issues deciphering things, but that can be overridden on a per-application basis, rather than needing to be packaged as the primary option.

SuperFluffy commented:
Is there some progress on this? I'd love to see this land.

syntaqx commented:
¯\_(ツ)_/¯ can't seem to get him to merge it.

VojtechVitek commented:
Open-source is tough, man. Merging this PR means I'd have to spend hours of testing the changeset in the pre-production and production environments with our projects.

I'm willing to do that, but I can't do that right away, since I have different priorities at this time.

Send bitcoin or ETH and I'll find some time sooner! :D

VojtechVitek commented:
You're right. I'm slow. And I'm sorry.

I'll try to test this when I find some time soonish.

syntaqx commented:
Happy 2019 @VojtechVitek - How're we doing on getting this merged friend? Especially given we have Semantic Versioning now, I feel like the issue is less complicated.

pzeinlinger commented:
Hey @VojtechVitek,
what about merging this PR? It's not complicated, and we have been running this for quite some time now in production. No issues at all.

FryDay commented:
Is this project no longer maintained? This is a pretty simple PR and has been open for 4 years.

VojtechVitek commented:
I'm sorry. I don't currently use go-chi/render.

This PR has some old module mgmt system, so I prefer not to merge it right away.

syntaqx commented:
I'm more than happy to include Go modules, assuming that you will actually make an effort to get this merged.

What are you using instead of this project, so that people can look to that?

Additionally, I'm more than happy to help maintain this project if you're open to it. It certainly needs a few modernization updates, but there's plenty of participation on the project that could use responses, and some additional features would make it a more than viable option for folks.and would be more than happy to even maintain the repository to provide it with some additional functionality that it would benefit from, b

syntaqx commented:
@VojtechVitek I've updated the pull request:

- Uses go modules
- Swapped Travis for GitHub Actions (the old tests relied on goimports, and given that I don't use Travis I just swapped to actions instead) - Let me know if you have any concerns here.
- Added a couple of everyone's favorite change: build badges in the readme

If we could move to getting this merged in, that would be awesome. Also, please let me know your thoughts on the future of this project, as I would be more than happy to provide it a bit of attention and some much needed updates, but I don't want another 4 years of PR delays ;)

syntaqx commented:
Sorry, couple minor updates there; didn't realize the linting wasn't passing. Actually good for review/merge now.

FryDay commented:
I would be willing to assist @syntaqx in maintaining Render. Chi and Render are certainly being used in production in various projects and really need to be kept current.

VojtechVitek commented on responder.go:
```diff
@@ -66,7 +66,7 @@ func PlainText(w http.ResponseWriter, r *http.Request, v string) {
 		w.WriteHeader(status)
 	}
-	w.Write([]byte(v))
+	w.Write([]byte(v)) //nolint:errcheck
```
nit: `_ = w.Write()` would be imho more idiomatic than using some special comments

VojtechVitek reviewed (approved):
LGTM

Thanks for switching to Go Modules and to Github Actions too. Great job!

For reference:
https://github.com/syntaqx/go-chi-render/runs/2451985236
https://github.com/syntaqx/go-chi-render/runs/2451985149
